src/backend/impl/src/repositories/proposal_repository.rs

```diff
@@ -0,0 +1,150 @@
+use super::{
+    memories::{init_proposals, ProposalMemory},
+    Proposal, ProposalId, ReviewPeriodState,
+};
+use backend_api::ApiError;
+use std::cell::RefCell;
+
+#[cfg_attr(test, mockall::automock)]
+pub trait ProposalRepository {
+    fn get_proposal_by_id(&self, proposal_id: &ProposalId) -> Option<Proposal>;
+
+    async fn create_proposal(&self, user_profile: Proposal) -> Result<ProposalId, ApiError>;
+
+    fn update_proposal(&self, proposal_id: &ProposalId, proposal: Proposal)
+        -> Result<(), ApiError>;
+}
+
+pub struct ProposalRepositoryImpl {}
+
+impl Default for ProposalRepositoryImpl {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ProposalRepository for ProposalRepositoryImpl {
+    fn get_proposal_by_id(&self, proposal_id: &ProposalId) -> Option<Proposal> {
+        STATE.with_borrow(|s| s.proposals.get(proposal_id))
+    }
+
+    async fn create_proposal(&self, proposal: Proposal) -> Result<ProposalId, ApiError> {
+        let proposal_id = ProposalId::new().await?;
+
+        STATE.with_borrow_mut(|s| {
+            s.proposals.insert(proposal_id, proposal);
+        });
+
+        Ok(proposal_id)
+    }
+
+    fn update_proposal(
+        &self,
+        proposal_id: &ProposalId,
+        proposal: Proposal,
+    ) -> Result<(), ApiError> {
+        self.get_proposal_by_id(proposal_id).ok_or_else(|| {
+            ApiError::not_found(&format!(
+                "Proposal with id {} not found",
+                proposal_id.to_string()
+            ))
+        })?;
+
+        // match (proposal.state, state.clone()) {
+        //     (ReviewPeriodState::InProgress, ReviewPeriodState::Completed) => {
+        //         proposal.state = state;
+
+        //         STATE.with_borrow_mut(|s| {
+        //             s.proposals.insert(proposal_id.clone(), proposal);
+
+        //             Ok(())
+        //         })
+        //     }
+        //     _ => Err(ApiError::invalid_argument(
+        //         "Invalid proposal state transition",
+        //     )),
+        // }
```

Comment:
@nathanosdev should we check the updated proposal validity here or in the service?

Reply:
I would do this types of "state machine" type things in the service and keep the repository a bit "dumb" in terms of business logic